### PR TITLE
Skip DECPS/MIDI output on Ctrl+C/Break

### PR DESCRIPTION
**Pull request by lhecker**
Silent MIDI notes can be used to seemingly deny a user's input for long
durations (multiple minutes). This commit improves the situation by ignoring
all DECPS sequences for a second when Ctrl+C/Ctrl+Break is pressed.

Additionally it fixes a regression introduced in 666c446:
When we close a tab we need to unblock/shutdown `MidiAudio` early,
so that `ConptyConnection::Close()` can run down as fast as possible.

## Validation Steps Performed
* In pwsh in Windows Terminal 1.16 run ``while ($True) { echo "`e[3;8;3,~" }``
  * Ctrl+C doesn't do anything ❎
  * Closing the tab doesn't do anything ❎
* With these modifications in Windows Terminal:
  * Ctrl+C stops the output ✅
  * Closing the tab completes instantly ✅
* With these modifications in OpenConsole:
  * Ctrl+C stops the output ✅
  * Closing the window completes instantly ✅

### CODE REVIEW

File: src/audio/midi/MidiAudio.cpp
Comment (lhecker) [inline src/audio/midi/MidiAudio.cpp:51]: This is the new trick.

File: src/audio/midi/MidiAudio.cpp
Comment (lhecker) [inline src/audio/midi/MidiAudio.cpp:55]: I'll explain reason for this change below in ControlCore.

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:409]: I wasn't sure where to put this...

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:418]: Is this the best way to create timers in Windows Terminal?

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:1518]: This is what required larger changes to `MidiAudio`:
On shutdown we have to tell the VT output thread to skip all DECPS/MIDI sequences. `_getMidiAudio` uses no mutex to allow thread-safe access of `_midiAudio` from multiple threads. But luckily we can either rely on how it's only called when the console lock is held, or we simply add another mutex there.

But both options are bad IMO: We had a long history of deadlock in Windows Terminal due to the main thread acquiring the console lock. And technically we don't need one, since all we want to do is call `BeginSkip` which is already inherently thread-safe. I simply dropped the lazy creation of `MidiAudio` which now allows me to blindly call `BeginSkip` witout if conditions and checks. I think this makes the code simpler and more robust. But it requires us to pass the `HWND` on all calls now for the internal lazy creation of DirectSound types.

Another downside is that I had to sprinkle the DirectSound headers into a bunch of .cpp files so that the destructor compiles without "undefined type" errors. I tried declaring a custom `MidiAudio::~MidiAudio` destructor, but that didn't make the compiler treat the destructor as an "extern" symbol either. I don't get why `std::unique_ptr` allows such type erasure where the compiler doesn't need to know the destructor definition, but `Microsoft::Wrl` and `wil::com_ptr` can't do that...

**Comment by DHowett** [comment]
@j4james FYI This might seem like a weird thing to do, but we got a report of the potential for DECPS blocking to be used as part of deceiving users into thinking that their session has hung (and they can't get out of it.)

**Comment by j4james** [comment]
OK I'll try and take a look at this over the weekend - it's going to take me a while to figure out how it works. It doesn't seem obvious to me at first glance how you can get away without locking the `MidiAudio` class.

But I do like the idea of being able to stop the sound with Ctrl+C.

**Comment by lhecker** [comment]
BTW the original issue mentions:
> The terminal’s sound buffer can store 16 notes of specified volume and duration. Additional sound controls are held in the communications input buffer invoking receive data flow control as needed.

Unfortunately I don't think it was ever specifically mentioned how the VT520 behaved. Did it block IO until a control sequence finished playing or did it allow up to 16 notes to be queued up in a small circular buffer?
Because I feel like DECPS would be _way_ more useful if it didn't block IO on every note. If you'd use it right now as - for instance - an indicator for a non-zero exit code, then it would always kinda freeze up your terminal briefly... 🤔

**Comment by j4james** [comment]
> Unfortunately I don't think it was ever specifically mentioned how the VT520 behaved. Did it block IO until a control sequence finished playing or did it allow up to 16 notes to be queued up in a small circular buffer?

As far we could establish on the VT525 that jerch tested, that 16 note buffer was not a thing. The `DECPS` sequence could only accept a single note, and it blocked until the note had finished playing. The WT implementation extends that to allow for multiple notes to be specified in a single `DECPS` sequence (as allowed for in the documentation), but otherwise matches the VT525 blocking behavior, i.e. we block until they've all finished playing.

These details were uncovered in the comment thread here: https://github.com/jerch/xterm.js/issues/1#issuecomment-966387953. It's really just the three messages at that point in the thread that are relevant to `DECPS` - we went on to discuss a bunch of unrelated audio stuff after that.

**Comment by j4james** [comment]
I've had a chance to play around with this PR, and try and figure out how it works without locking, i.e. what is keeping the `StateMachine`, `Renderer`, etc. open until it's safe to shutdown. So what I did was place some `OutputDebugString` calls in the constructors and destructors of various classes (`TermControl`, `StateMachine`, `Renderer`, etc.).

When I open a tab, I see the constructors being called, but when I close a tab, nothing. Keep opening and closing tabs - lots of construction, but no destruction. So I close the whole terminal and suddenly there's a long string of destructor calls to free up all those tabs that had previously been used.

So it seems the trick is that we're just not releasing any memory until the whole app is closed? That's surely not intentional is it? If anyone is using the terminal for an extended period of time it's just going to keep using more and more memory. Or am I misunderstanding something here?

And in the case of conhost, nothing ever gets destroyed as far I can tell. I suppose that is OK when the whole app is exiting anyway, but I'd like to know if that was an intentional optimization rather than a bug that we're just ignoring.

Anyway I'm assuming this change in behavior was introduced at some point in the past, so it's not technically a problem with this PR, but I'm not certain this PR would work without that change being in place.

**Comment by lhecker** [comment]
@j4james I fixed that bug in #14228. Thanks for pointing it out! 😊

**Comment by lhecker** [comment]
> And in the case of conhost, nothing ever gets destroyed as far I can tell. I suppose that is OK when the whole app is exiting anyway, but I'd like to know if that was an intentional optimization rather than a bug that we're just ignoring.

It's intentional. We use `ExitProcess()` in `ServiceLocator::RundownAndExit` to exit immediately. I tried using `exit()` once to do a clean exit, but that didn't work at all. The primary problem is the heavy reliance on the `Globals` object and everything it ever touched to always exist. So for instance at one point I fixed a random crash that was caused by the `CursorBlinker`, because it kept blinking even when we were in the process of shutting down and the `Renderer` was already a `nullptr`. The issue is so bad that we had to add the `#ifndef NDEBUG` here: https://github.com/microsoft/terminal/blob/1f19ed0cd2e04dd46852300c369960d6fa69a8fa/src/interactivity/win32/window.cpp#L68-L81

**Comment by j4james** [comment]
OK, with PR #14228 applied, I've taken another look at this, and I think I understand how it works now. In case anyone else shared my confusion, the reason we don't need a lock is because the `ConptyConnection::Close` method [waits for the output thread to exit](https://github.com/microsoft/terminal/blob/274bdb31da9799fe5e2da1cb1d5a5ac60b3c6a51/src/cascadia/TerminalConnection/ConptyConnection.cpp#L571-L574), and that `Close` is now called from the `ControlCore::Close` (I think since PR #13882), so there should be no risk of anything being destroyed while the midi code is still finishing off.

With that understood, and the understanding that conhost is just going to exit immediately, I'm happy with this approach. I suspect there might be weird edge cases where we could get a bit of a delay in closing, but I think the simpler implementation more than makes up for that. Also I love the Ctrl+C functionality.

I have a few minor nits on the PR which I'll comment on later, but I don't think there's anything blocking.

File: src/terminal/adapter/DispatchTypes.hpp
Comment (j4james) [inline src/terminal/adapter/DispatchTypes.hpp:9]: Nit: I think the duration parameters would be better left as they were with the original microsecond precision, and then use an explicit `duration_cast` to get the milliseconds when you need them. Maybe one day we might even be able to use the exact duration value for the wait. As it is, we're left with a `count()` of an unknown scale, which is kind of meaningless. 

Also having to use `::Microsoft::Console::VirtualTerminal::MidiDuration` all over the place is bit painful.

File: src/audio/midi/MidiAudio.cpp
Comment (j4james) [inline src/audio/midi/MidiAudio.cpp:83]: This comment could do with an update. Maybe something like "By waiting on the skip event...", and possibly also mention that it applies to both shutdown and skipping notes.

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (j4james) [inline src/cascadia/TerminalControl/ControlCore.cpp:1424]: With the code above removed, this comment doesn't make sense. Maybe just need to drop the "then".

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (j4james) [inline src/cascadia/TerminalControl/ControlCore.cpp:1412]: Again this comment doesn't make sense with the Unlock code removed. Maybe just drop everything after "reacquire the terminal lock".

File: src/host/outputStream.cpp
Comment (j4james) [inline src/host/outputStream.cpp:377]: Same deal as the comments in `ControlCore::_terminalPlayMidiNote`. Just drop the "then".

File: src/host/outputStream.cpp
Comment (j4james) [inline src/host/outputStream.cpp:391]: Same as `ControlCore::_terminalPlayMidiNote` again. Just drop everything after "reacquire the console lock".

File: src/audio/midi/MidiAudio.cpp
Comment (j4james) [inline src/audio/midi/MidiAudio.cpp:80]: Since we aren't using this exception anymore, we should get rid of the definition in [`statemachine.hpp`](https://github.com/microsoft/terminal/blob/f2b361c1463371ac2be681235470343ae48c23b5/src/terminal/parser/stateMachine.hpp#L71-L76), and the other bits of code associated with it that were added in PR #13208. There was a `catch` that was added to [`ControlCore::_connectionOutputHandler`](https://github.com/microsoft/terminal/blob/07201d6cd12870755efa7bb52e220b0b10102730/src/cascadia/TerminalControl/ControlCore.cpp#L1717-L1721) and another in  [`StateMachine::_SafeExecute`](https://github.com/microsoft/terminal/blob/f2b361c1463371ac2be681235470343ae48c23b5/src/terminal/parser/stateMachine.cpp#L2041-L2044). You may also need to add back a bunch of noexcepts in the `StateMachine` class if the audit complains.

**Review by j4james — APPROVED** [review]
Looks good to me. Just some comment updates, and clean up of obsolete code. I don't like the `MidiDuration` thing, but feel free to ignore that feedback if you disagree.

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/ControlCore.cpp:409]: Hmm... You want this behavior to be a part of the terminal handling it right? Like, if a user binds an action to ctrl+c, the action should be executed and this new behavior shouldn't happen right?

If that's the case, I suggest taking a look at `ControlCore::TrySendKeyEvent()`. By that point, we know that the input wasn't handled by an action so we can intercept it right before it goes to the shell (NOTE: this is where we have logic to update the selection from the keyboard when not in mark mode).

**Comment by DHowett** [comment]
> Like, if a user binds an action to ctrl+c, the action should be executed and this new behavior shouldn't happen right?

Yeah, unfortunately this is correct. It should be overridable by an action just like _actual `^C`_ isz.

File: src/terminal/adapter/DispatchTypes.hpp
Comment (lhecker) [reply src/terminal/adapter/DispatchTypes.hpp:9]: I don't mind changing back to `microsecond` at all, but generally I'd prefer having the `MidiDuration` type alias. It makes not just discovering where this type is used easier, but also what the actual type is - similar to `VTInt`.

File: src/host/server.h
Comment (lhecker) [reply src/host/server.h:159]: Actually I removed the `unique_ptr` to simplify destruction. The reasoning behind this is annoying and difficult and explained in detail in this comment: https://github.com/microsoft/terminal/pull/14214#discussion_r995126594

But the gist of it is: In order to correctly shut down without race conditions we either need to protect access to `_midiAudio` with a mutex or remove the `unique_ptr`. The latter is the safer alternative.

File: src/host/outputStream.cpp
Comment (lhecker) [reply src/host/outputStream.cpp:371]: I personally disagree (but I'm open to changing my opinion).
I feel like there's times were specific types/units are warranted - for instance if you want to represent an exact 32-Bit integer you should use `uint32_t` - but there are also times were obfuscating the underlying type makes coding easier. Most programming languages for instance don't even have multiple duration types like C++. Instead they simply got a single `Duration` type or similar, which contains nanoseconds for instance. But even with C++ this is partially the same deal: `std::chrono::duration`'s underlying "ratio" is supposed to be transparent which is evident by how you can transparently convert any duration into any other duration and the copy operator will multiply/divide by the given underlying ratios.

Additionally, in this particularly case the duration we want to represent is neither in milliseconds, nor in microseconds at all: It's unit is in 1/32th of a second and we could choose to represent it in nanoseconds. Since `std::chrono::duration` obfuscates the unit of durations and allows duration conversions, saying that it's in `milliseconds` is kind of redundant IMO. It would make more sense in my opinion if the argument was a primitive integer for instance that didn't had the ratio builtin like the duration type does.

File: src/terminal/adapter/DispatchTypes.hpp
Comment (j4james) [reply src/terminal/adapter/DispatchTypes.hpp:9]: I can understand that for something like `VTInt`, where changing from say `size_t` to `int` is still semantically the same thing, but that's not the case here. You're not really changing the type - you're changing the scale. While `size_t 123` and `int 123` are essentially equivalent, `123 milliseconds` and `123 microseconds` are definitely not!

File: src/host/outputStream.cpp
Comment (j4james) [reply src/host/outputStream.cpp:371]: I just want to make a point about the "1/32th of a second" thing. That's a limit of the `DECPS` sequence in particular, but my intention was to leave the API open to potentially being used for other operations as well. I chose microseconds as the unit because that seemed like a reasonable generic scale that would likely cover most use cases.

File: src/terminal/adapter/DispatchTypes.hpp
Comment (j4james) [reply src/terminal/adapter/DispatchTypes.hpp:9]: I should add that I wouldn't mind so much if you included a `duration_cast` where you're doing the `count()`, because that's mitigates the risk of the code breaking if someone redefines the `MidiDuration` type.

Comment (lhecker) [reply src/terminal/adapter/DispatchTypes.hpp:9]: > I can understand that for something like `VTInt`, where changing from say `size_t` to `int` is still semantically the same thing, but that's not the case here. You're not really changing the type - you're changing the scale. While `size_t 123` and `int 123` are essentially equivalent, `123 milliseconds` and `123 microseconds` are definitely not!

I'm fine with reverting all those parts of the PR. 🙂
But while I do that I do have a question: We can represent the DECPS duration as a custom
```cpp
std::chrono::duration<VTInt, std::ratio<1, 32>>
```
...but should we really have all those functions spell out that type every time? Wouldn't a `DECPSDuration` (or `MidiDuration`) alias better?

> I should add that I wouldn't mind so much if you included a `duration_cast` where you're doing the `count()`, because that's mitigates the risk of the code breaking if someone redefines the `MidiDuration` type.

I actually intentionally used an explicit `std::chrono::milliseconds` as the parameter type for the `MidiAudio` class because of that. Even when I revert those changes now I'd like to keep that part as I feel like the `MidiAudio` class isn't really part of the VT system and so it doesn't have to use the same parameter type IMO... I guess? 🤔

Comment (DHowett) [reply src/terminal/adapter/DispatchTypes.hpp:9]: (James' explanation for why he wants MIDI durations to be separate from DECPS durations is sound to me; I suspect you just haven't seen it yet!)

Comment (j4james) [reply src/terminal/adapter/DispatchTypes.hpp:9]: > I actually intentionally used an explicit `std::chrono::milliseconds` as the parameter type for the `MidiAudio` class because of that.

Oh sorry. I completely missed that. The lack of cast is fine then. I still don't like the `MidiDuration`, but maybe that's just a personal preference.

> `MidiAudio` class isn't really part of the VT system and so it doesn't have to use the same parameter type IMO

I agree completely. But as I said below, I chose microseconds for the type because I thought it was a reasonable scale that would cover most use cases. The fact that the current wait operation takes milliseconds is more of an implementation detail and shouldn't really be deciding the API parameters. This is not a big deal though. I just wanted to make my case. I'll leave it up to you now.

**Comment by lhecker** [comment]
> Yeah, unfortunately this is correct. It should be overridable by an action just like _actual `^C`_ is.

@DHowett @carlos-zamora Since a handled action returns `true` in the key event handler the character event handler is never called. So for instance if I run a DECPS sequence and select some text, pressing Ctrl+C won't cancel the sequence. I think the character event handler is a better fit than the key event one since C and Break have two different virtual keys and the behavior is a bit awkward...
* Break is vkey `0x13` (`VK_PAUSE`)
* Ctrl+Break is `0x03` (`VK_CANCEL`)
* but Ctrl+C is of course not `VK_CANCEL`, it's `0x43`, aka `'C'` 😕

On the other hand in the character handler it's just a single `ch == L'\x3'`.

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:1427]: nit: `_owningHwnd` will change at runtime in the future; is MidiAudio prepared to handle that?

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (DHowett) [reply src/cascadia/TerminalControl/ControlCore.cpp:1515]: This is OK for now. We can reevalute the header situation in the future. :)

**Review by DHowett — APPROVED** [review]
I'm cool with this. Thank you!

**Comment by DHowett** [comment]
@msftbot merge this in 15 minutes

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (lhecker) [reply src/cascadia/TerminalControl/ControlCore.cpp:418]: The documentation isn't particularly great
> ## DispatcherQueueTimer.Start Method
> ### Definition
> Starts the timer.

However wading through the underlying code for this, the timer is canceled before it gets restarted.

File: src/cascadia/TerminalControl/ControlCore.cpp
Comment (lhecker) [reply src/cascadia/TerminalControl/ControlCore.cpp:1427]: The device will be reinitialized if the HWND changes. But we don't expect this to happen too often right?

Comment (DHowett) [reply src/cascadia/TerminalControl/ControlCore.cpp:1427]: You're right, we don't!

**Comment by DHowett** [comment]
I'm admin-merging with one Team sign-off since @j4james is the local area expert & I trust his review here :smile:

Thanks!

File: src/host/output.cpp
Comment (lhecker) [inline src/host/output.cpp:515]: FYI reading another PR, I just realized that this change isn't quite correct. This should still call `midiAudio.BeginSkip()`, just like `ControlCore` does on shutdown.

**Pull request merged**
